Midi 82/vel transformer blogpost

WojciechMat opened the pull request:
Initial version of a blog post without final sections. 
I was unable to change width of a audio controller, so I can fit only two columns with samples in a row.
Let me know your thoughts and suggestions :))

WojciechMat committed:
initial blogpost with no demonstration

WojciechMat committed:
-

WojciechMat committed:
quantization sample with two columns

WojciechMat committed:
second row with quantization samples and titles for samples

WojciechMat committed:
samples and pieces

WojciechMat committed:
transformer scheme

WojciechMat committed:
different place for transformer.png

WojciechMat committed:
table with hyperparemeters

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,241 @@
+  {% algrtmImg MIDI-velocity-transformer/samples/9115-real.png pianoroll 170px %}
+  {% algrtmAudio MIDI-velocity-transformer/samples/9115-real.mp3 %}
+        
+  {% algrtmImg MIDI-velocity-transformer/samples/9115-pred-untrained.png pianoroll 170px %}
+  {% algrtmAudio MIDI-velocity-transformer/samples/9115-pred-untrained.mp3 %}
```
Missing files 🚨

WojciechMat replied:
should be ok now :+1:

WojciechMat committed:
missing files

WojciechMat committed:
delete unused files

WojciechMat committed:
conclusion

WojciechMat committed:
more text

WojciechMat committed:
contact???

WojciechMat commented:
I have added some more text about tokenization as well as conclusion, references and missing file from before. Everything should work now. I am not sure as to contact info at the end of the post, maybe you know how you would like for it to look like? 
Would love to hear your thoughts on what more to include, what to change in the text, if I should include more clarifications or make the content more engaging let me know! :fire:

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,266 @@
+---
+title: MIDI Velocity Prediction with Transformer
```
```suggestion
title: Modelling dynamic expression in piano performance
```

I think we shouldn't mention MIDI or velocity in the title, as it's to technical - my suggestions aims to provide broader goal for what we're doing. I think you can add "with Transformers" if you like

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,266 @@
+
+{% algrtmImgBanner MIDI-velocity-transformer/banner-1.jpg pianoroll %}
+
+MIDI velocity is a crucial element in music dynamics, determining the force with which a note is played, 
+which profoundly influences the emotional quality of music. 
```
I think you should try to explain the problem without referencing MIDI, and then you could introduce MIDI as a data structure related to this problem 🤔

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,266 @@
+MIDI velocity is a crucial element in music dynamics, determining the force with which a note is played, 
+which profoundly influences the emotional quality of music. 
+
+If you were to take a sequence of notes and predict their velocities by an untrained model, this is what you wold have ended up with:
```
Those samples could be a good demonstration of what velocity is and how it affects the music - you could say that the velocities were randomized here.

Saying it's generated by an untrained model is confusing here, because you haven't introduce the problem that the model has to solve yet.

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,266 @@
+within quantized MIDI data.
+
+
+### Model Overview
```
I think we don't need a model overview at all here. We can assume that the readers know what a transformer is, and we can include some references in the text.

What this and the next section is missing for me, is a description of how we convert the piano performance problem into a list of tokens problem.

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,266 @@
+mark a significant stride toward the realm of emotionally resonant AI music generation.
+### Data Preprocessing
+#### MIDI data
+MIDI data describes notes by 5 features:
+   1. Pitch - Represented as a number between 0 and 127 (or 21 to 108 for piano keys, reflecting the standard 88-key keyboard).
+   2. Start - Indicates the moment a key is pressed, measured in seconds.
+   3. End - Marks the second when the key is released.
+   4. Duration - calculated as the time elapsed between the key's press and release.
+   5. Velocity - ranging from 0 to 128, indicating the intensity of the key press.
```
List of points is not the best format to describe a data structure. One alternative would be to make a code snippet showing a "note" class design, and referring to piano performance as a list of notes:

```python
class Note:
    pitch: int
    start: float
    end: float
    velocity: int
```

You can also use https://mermaid.live/ to make a class diagram, I think our blog should support it just like github:

```mermaid
classDiagram
    class Note{
      pitch: int
      velocity: int
      start: float
      end: float
    }
```

(this does not look great, but you could play around with it)

WojciechMat replied:
I do not think it does support mermaid :(

roszcz replied:
I just updated `master`, and now it does 🎉

WojciechMat replied:
:fire::fire::fire:

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,266 @@
+
+</div>
+
+### Model Architecture
+{% algrtmImgBanner MIDI-velocity-transformer/transformer.png transformer%}
+A transformer built as described in [Attention is all you need](https://arxiv.org/abs/1706.03762) paper was used for this task.
+The important hyperparameters:
+| hyperparameter | number |
+| -------------- | :-----: |
+| Number of layers in encoder and decoder | **6** |
+| Nuber of heads in attention layers | **8** |
+| Dimension of encoder and decoder outputs | **512** |
+| Dimension of a hidden layer of position-wise fast-forward network from each layer of encoder and decoder | **2048** |
```
Like I mentioned, we don't need to explain what transformers are. I think the only relevant information we should share about our architecture is the number of trainable parameters.

Also, I'm agains using diagrams from external sources, either make your own, or just link to the source :)

roszcz commented on blog/source/_posts/MIDI-velocity-transformer.md:
```diff
@@ -0,0 +1,266 @@
+#### Data
+The vocabulary for our model was generated by considering every combination of quantized information from MIDI notes. The source vocabulary consisted of 795 unique values, while the target vocabulary included 131 values, accounting for start, end, and padding tokens.
+
+Our training dataset comprised approximately 200 hours of musical data sourced from the 
+[roszcz/maestro-v1](https://huggingface.co/datasets/roszcz/maestro-v1) dataset, which includes 1276 pieces of classical music performed during piano competitions. Each musical piece was segmented into 128-note sequences, with a 64-note overlap between adjacent samples. These sequences were quantized, and each note was mapped to its corresponding index in the source and target vocabularies.
```
I'm not the owner of maestro, you should cite the original source: https://magenta.tensorflow.org/datasets/maestro

WojciechMat committed:
random velocities, different title, introduction initial draft

WojciechMat committed:
class diagram

WojciechMat committed:
Merge branch 'master' into MIDI-82/vel-transformer-blogpost

WojciechMat committed:
merge master

WojciechMat committed:
updated blog

WojciechMat committed:
adjustments

WojciechMat commented:
Mermaid does work, thanks :)),
I've added notes dataframe representation, some code snippet, removed redundant info on transformer and it's architecture.
I  invite you to read, suggest changes and modify the text too.

roszcz committed:
pin node version

WojciechMat committed:
change text, delete unused banner

WojciechMat committed:
fix pieces, add chopin from maestr-v1-sustain

WojciechMat committed:
Merge branch 'MIDI-82/vel-transformer-blogpost' of github.com:Nospoko/blog into MIDI-82/vel-transformer-blogpost

WojciechMat committed:
blog update

WojciechMat committed:
midi desc

WojciechMat committed:
update - tiny changes

WojciechMat committed:
FEED forward network fix
